cmd/outline-ss-server/metrics.go

```diff
@@ -49,9 +62,106 @@ type outlineMetrics struct {
 var _ service.TCPMetrics = (*outlineMetrics)(nil)
 var _ service.UDPMetrics = (*outlineMetrics)(nil)
 
+type ReportTunnelTimeFunc func(IPKey, ipinfo.IPInfo, time.Duration)
+
+type activeClient struct {
+	mu              sync.Mutex
+	IPKey           IPKey
+	clientInfo      ipinfo.IPInfo
+	connectionCount int
+	startTime       time.Time
+}
+
+type IPKey struct {
+	ip        netip.Addr
+	accessKey string
+}
+
+type tunnelTimeTracker struct {
+	mu               sync.Mutex
+	activeClients    map[IPKey]*activeClient
+	reportTunnelTime ReportTunnelTimeFunc
+}
+
+// Reports time connected for all active clients, called at a regular interval.
+func (t *tunnelTimeTracker) reportAll(now time.Time) {
+	if len(t.activeClients) == 0 {
+		logger.Debugf("No active clients. No TunnelTime activity to report.")
+		return
+	}
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	for _, c := range t.activeClients {
+		t.reportDuration(c, now)
+	}
+}
+
+// Reports time connected for a given active client.
+func (t *tunnelTimeTracker) reportDuration(c *activeClient, now time.Time) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	connDuration := now.Sub(c.startTime)
+	logger.Debugf("Reporting activity for key `%v`, duration: %v", c.IPKey.accessKey, connDuration)
+	t.reportTunnelTime(c.IPKey, c.clientInfo, connDuration)
+	// Reset the start time now that it's been reported.
+	c.startTime = Now()
+}
+
+// Registers a new active connection for a client [net.Addr] and access key.
+func (t *tunnelTimeTracker) startConnection(clientInfo ipinfo.IPInfo, clientAddr net.Addr, accessKey string) {
+	hostname, _, _ := net.SplitHostPort(clientAddr.String())
+	ipKey := IPKey{ip: netip.MustParseAddr(hostname), accessKey: accessKey}
+
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	c, exists := t.activeClients[ipKey]
+	if !exists {
+		c = &activeClient{
+			IPKey:      ipKey,
+			clientInfo: clientInfo,
+			startTime:  Now(),
+		}
+	}
+	c.connectionCount++
+	t.activeClients[ipKey] = c
+}
+
+// Removes an active connection for a client [net.Addr] and access key.
+func (t *tunnelTimeTracker) stopConnection(clientAddr net.Addr, accessKey string) {
+	hostname, _, _ := net.SplitHostPort(clientAddr.String())
+	ipKey := IPKey{ip: netip.MustParseAddr(hostname), accessKey: accessKey}
+
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	c, exists := t.activeClients[ipKey]
+	if !exists {
+		logger.Warningf("Failed to find active client")
+		return
+	}
+	c.mu.Lock()
+	c.connectionCount--
+	c.mu.Unlock()
+	if c.connectionCount <= 0 {
+		t.reportDuration(c, Now())
+		delete(t.activeClients, ipKey)
+		return
+	}
+}
+
+func newTunnelTimeTracker(report ReportTunnelTimeFunc) *tunnelTimeTracker {
+	tracker := &tunnelTimeTracker{activeClients: make(map[IPKey]*activeClient), reportTunnelTime: report}
+	ticker := time.NewTicker(tunnelTimeTrackerReportingInterval)
+	go func() {
```

Comment:
It's possible to generate the metrics on demand, on scrape, with https://pkg.go.dev/github.com/prometheus/client_golang/prometheus#Collector. See the example.

That may be a better approach than having the timed loop, because the scraper can determine the frequency at which it's collecting the data, and the metrics will all be in sync. It also saves work if there's only one scraper (which is usually the case).

I think the way it is is fine, so it's up to you.